### ISSUE TITLE
Guide on how this project works in detail for beginners

### ISSUE THREAD

**Issue by danielstorch** [reactGo/reactGo#132]
Hey,

so im trying to understand this project in detail and have some issues. After installing mongodb and cloning the project, i had to use `npm install` twice before the project was working. I'm still not sure what the issue was.

The problem i have now is that i don't understand the whole user identification. How do i add a user so i can Log in ?
Is there any documentation on how the authentication works?
What is the `secret.js` file used for and what do i have too change there? 
Is there any other configurations that have to be done on first setup?

I'm sorry if this is documented some where. I'm quite new too all this and have some difficulties understanding how everything is put together.

**Comment by choonkending** [comment]
@danielstorch Thanks for raising this issue.

Interesting, I don't think I've had to `npm install` twice in order for things to work. It's likely that your connection dropped out? Which version of `node` are you using? You can type `node -v` in your command line to check.

- How do i add a user so i can Log in ?
Currently there is a server endpoint that handles it without the front-end code being hooked up to handle it. So you have to paste the following code into your browser console.
```javascript
var http = new XMLHttpRequest();
var url = "http://localhost:3000/signup";
var params = "email=example@ninja.com&password=ninja";
http.open("POST", url, true);http.setRequestHeader("Content-type", "application/x-www-form-urlencoded");

http.send(params);
```
You can replace the email and password with your own email and password.
I will look into hooking up the proper auth flow for registration as well so it will not cause as many problems to new users. Issue raised https://github.com/choonkending/react-webpack-node/issues/129

- Is there any documentation on how the authentication works?

Currently there is none. We use a middleware called [PassportJS](http://passportjs.org/) that is pretty sweet in handling authentication. 
- What is the secret.js file used for and what do i have too change there?

The secret file is where you store your credentials for any sort of service. For example, if you were to use Google Authentication methods, you would be given a client ID and secret key, *which you should never reveal to anyone*. The secret file should ideally be stored somewhere safe and not committed to GitHub. The reason why I pushed it up is for users to see how I defined it. I am by no means an expert in configuration and security, so it would be good if anyone raised an issue if there is a better way of handling this.
- Is there any other configurations that have to be done on first setup?

If you wish to test Google Auth locally or on Heroku, you might want to read [this](https://github.com/choonkending/react-webpack-node#questions). Besides MongoDB, I do not think anything else requires configuration. 

These are all valid questions that should be added into the README! It might be high time for me to make the README more reader-friendly hehe.

**Comment by danielstorch** [comment]
Thanks for the fast and detailed replay, this is an enormous help.

- `npm install` error

The connection drop out might be it. I had several ` Error: connect ECONNREFUSED`.

- secret.js

So if i don't use any external services, i don't have to put there anything? 
What about the DB on Heroku? is there anything else than installing the mongoDB addon on Heroku i have to do?

When trying to Login with an non existent email and password, the server crashes with this stack trace
```
serialize
_http_outgoing.js:345
    throw new Error('Can\'t set headers after they are sent.');
    ^

Error: Can't set headers after they are sent.
    at ServerResponse.OutgoingMessage.setHeader (_http_outgoing.js:345:11)
    at ServerResponse.header (/Users/danielstorch/Documents/Develop/react-webpack-node-master/node_modules/express/lib/response.js:718:10)
    at ServerResponse.send (/Users/danielstorch/Documents/Develop/react-webpack-node-master/node_modules/express/lib/response.js:163:12)
    at /Users/danielstorch/Documents/Develop/react-webpack-node-master/server/controllers/users.js:19:21
    at allFailed (/Users/danielstorch/Documents/Develop/react-webpack-node-master/node_modules/passport/lib/middleware/authenticate.js:87:18)
    at attempt (/Users/danielstorch/Documents/Develop/react-webpack-node-master/node_modules/passport/lib/middleware/authenticate.js:160:28)
    at Strategy.strategy.fail (/Users/danielstorch/Documents/Develop/react-webpack-node-master/node_modules/passport/lib/middleware/authenticate.js:277:9)
    at verified (/Users/danielstorch/Documents/Develop/react-webpack-node-master/node_modules/passport-local/lib/strategy.js:82:30)
    at Query.<anonymous> (/Users/danielstorch/Documents/Develop/react-webpack-node-master/server/config/passport/local.js:18:22)
    at /Users/danielstorch/Documents/Develop/react-webpack-node-master/node_modules/mongoose/node_modules/kareem/index.js:177:19
    at /Users/danielstorch/Documents/Develop/react-webpack-node-master/node_modules/mongoose/node_modules/kareem/index.js:109:16
    at nextTickCallbackWith0Args (node.js:433:9)
    at process._tickCallback (node.js:362:13)

```

**Comment by choonkending** [comment]
> The connection drop out might be it. I had several Error: connect ECONNREFUSED.

It might be because you are not running `mongod`? If you clone this boilerplate as-is, you will need to set up mongodb in your environment. Read more [here](https://github.com/choonkending/react-webpack-node#setting-up-your-database). However, it is relatively easy to strip out the mongoDB dependency, give it a go, and if you have any problems give a shout here.

> So if i don't use any external services, i don't have to put there anything? 

If you don't, as long as your server code does not reference that variable in the secrets file, you're golden. :)

> What about the DB on Heroku? is there anything else than installing the mongoDB addon on Heroku i have to do?

You would still have to register the login on the DB instance on Heroku. So you'd still have to run this in your browser console
```bash

// REPLACE url with your heroku endpoint
var http = new XMLHttpRequest();
var url = "https://www.react-webpack-node.herokuapp.com/signup";
var params = "email=example@ninja.com&password=ninja";
http.open("POST", url, true);http.setRequestHeader("Content-type", "application/x-www-form-urlencoded");

http.send(params);
```

However, this is a temporary solution. I am aware that the lack of a registration flow is tripping new users when they begin exploring this repo. https://github.com/choonkending/react-webpack-node/issues/129

I'll look into fixing it sometime this week.

**Comment by danielstorch** [comment]
Perfect, everything running fine at the moment. I really appreciate your help.

Did you look into the stack trace i posted? When trying to Login with a use which does not exist, the server crashes.

**Comment by choonkending** [comment]
Yep I saw that. Part of what I'm planning to fix this week when I get a
breather hehe :)

**Comment by danielstorch** [comment]
Ok thanks and sorry for bombing you with question, because here is the next one :P.

 Im trying to make a simple get request and can't get it to work. I postet the question on [stackoverflow](http://stackoverflow.com/questions/35280690/simple-get-request-with-node-js-and-express). Maybe you have the answer for it?

**Comment by choonkending** [comment]
@danielstorch It seems someone has already answered your stackoverflow question. :smile: 

> Also, where is the best place to load data from a user which newly logged in ? Like a onetime loading data for a user.

Can you provide a bit more context (or a use case) so I can better understand what you're trying to solve?

**Comment by danielstorch** [comment]
I will try to explain. When a user manually logs in, i want to check for initial data for this user. But only ONCE.
The same should happen if the user leaves the website but was still logged in and returns (session). I want to check for initial data once. 
Right now i slowed it by making an request in the `Login component` and  `navigation component` 
checking for `if(authenticated)`.

**Comment by danielstorch** [comment]
Hey, i think i found out what causes the crash of the server when trying to login with a not valid use.
`res.status(200).send('OK');` gets send before  `if(err) return next(err);`. Which than causes the error `Error: Can't set headers after they are sent.`
```
exports.postLogin = function(req, res, next) {
  // Do email and password validation for the server
  passport.authenticate('local', function(err, user, info) {
    if(err) return next(err);
    if(!user) {
      req.flash('errors', {msg: info.message});
    }
    // Passport exposes a login() function on req (also aliased as logIn()) that can be used to establish a login session
    req.logIn(user, function(err) {
      if(err) return next(err);
    });
     res.status(200).send('OK');
  })(req, res, next);
};
```

This works for me without crash:

```
exports.postLogin = function(req, res, next) {
  // Do email and password validation for the server
  passport.authenticate('local', function(err, user, info) {
    if(err) return next(err);
    if(!user) {
      req.flash('errors', {msg: info.message});
    }
    // Passport exposes a login() function on req (also aliased as logIn()) that can be used to establish a login session
    req.logIn(user, function(err) {
      if(err) return next(err);
      res.status(200).send('OK');
    });
  })(req, res, next);
};
```

**Comment by choonkending** [comment]
Hey thanks for that. I'll be getting onto this issue once I finish and bump
up the version for the previous PR (I think 138). Once that's done, this
will be a priority to fix + document. :)

**Comment by choonkending** [comment]
What a derp btw haha

**Comment by choonkending** [comment]
@danielstorch https://github.com/choonkending/react-webpack-node/commit/be1bb3ebdd29e457d0cd89dec98ce29397e4566f fixes the bug (thanks to your finding), but I'm still in the process of fixing tests + fixing up the flow.

**Comment by mikeon** [comment]
@choonkending If it's no problem to ask here, can you give a hint how to force server side rendering to allow use external library like react-media-player? If I create some components which use this library I cannot compile project, the error I get: 
`public\assets\server.js:33508
            return document.documentElement[curr] ? curr : prev;`
I tried to use check `if (typeof windows !== 'undefined'` before use component but it still something wrong. 
Thanks for any reply

**Comment by slavab89** [comment]
This is a pretty old issue and we're currently in a process of cleaning up the project's issues.
Regarding your use case, `document` is not present on the server side so if that library requires it then you'll have to disable it on the server side.
If you still experience issues with this, feel free to comment here or open a new issue.
